compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1609,44 +1609,66 @@ pub type PlaceholderType = Placeholder<BoundVar>;
 pub type PlaceholderConst = Placeholder<BoundVar>;
 
 /// A `DefId` which is potentially bundled with its corresponding generic parameter
-/// in case `did` is a const argument.
-///
-/// This is used to prevent cycle errors during typeck
-/// as `type_of(const_arg)` depends on `typeck(owning_body)`
-/// which once again requires the type of its generic arguments.
-///
-/// Luckily we only need to deal with const arguments once we
-/// know their corresponding parameters. We (ab)use this by
-/// calling `type_of(param_did)` for these arguments.
+/// in case `did` is a const argument. Used to prevent cycle errors during `typeck`.
 ///
 /// ```rust
 /// #![feature(const_generics)]
 ///
 /// struct A;
 /// impl A {
-///     fn foo<const N: usize>(&self) -> usize { N }
+///     fn foo<const N: usize>(&self) -> [u8; N] { [0; N] }
+///     //           ^ const parameter
 /// }
 /// struct B;
 /// impl B {
-///     fn foo<const N: u8>(&self) -> usize { 42 }
+///     fn foo<const M: u8>(&self) -> usize { 42 }
+///     //           ^ const parameter
 /// }
 ///
 /// fn main() {
 ///     let a = A;
-///     a.foo::<7>();
+///     let _b = a.foo::<7>();
+///     //               ^ const argument
 /// }
 /// ```
+///
+/// The important section here is the call `a.foo::<7>()`. We do not know which
+/// `foo` is used until we know the type of `a`.
+///
+/// We only know the type of `a` once we are type`typeck(main)`.
+/// At that point we also want to know the type of `_b` which depends on the
+/// the value of the const argument.
+///
+/// To evaluate that const argument we need to know its type,
+/// which we would get using `type_of(const_arg)`. This requires us to
+/// resolve `foo` as it can be either `usize` or `u8` in this example.
+/// However, resolving `foo` once again requires `typeck(main)` to get the type of `a`,
+/// which results in a cycle.
+///
+/// We only care about the const argument once we know which paramter they instantiate.
+/// So instead of using `type_of(const_arg)` we can use `type_of(const_param)` which does
+/// not depend on `typeck`.
+///
+/// This however requires us to pass the `const_param` to every place 
+/// which potentionally needs the type of a const argument and gets
+/// used during `typeck`. An example might be `tcx.typeck(const_arg)`,
+/// which we instead replace with `tcx.typeck_const_arg(const_arg, const_param)`.
+///
+/// As we share the same code paths between const args and const params we
```

Comment:
I have no idea what this means... args and params are totally different things, how could any code be shared?

Reply:
yeah, that sentence is incorrect. It should be "As we share the same code paths between const args and other constants"